library/pod-security-policy/capabilities/template.yaml

```diff
@@ -16,10 +19,12 @@ spec:
           properties:
             allowedCapabilities:
               type: array
+              description: "A list of Linux capabilities that can be added to a container."
```

Comment:
In general, I think background conceptual context should go in the constraint's description annotation, since that is where it would read most naturally in a markdown doc.

For people accessing the fields via `kubectl explain`, we should assume that they have seen the top-level `DESCRIPTION` output.

That adds a second note though...

`metadata.annnotations.description` has two key flaws:

1. The name is wrong: annotations should be namespaced to gatekeeper.sh
2. The annotation doesn't show up with `kubectl explain`

And one benefit, namely that annotations come at the top of the YAML

We should really be doing the following:

1. Putting what we are currently putting in `metadata.annotations.description` in the description for the top-level field of the CT's jsonschema (`spec.crd.spec.validation.openAPIV3Schema.description`)
2. When generating a CRD from the CT, using that root `description` field, as the root `description` field for the CRD's schema

That will work with `kubectl explain`

Unfortunately, I don't want to just get rid of the current `description` annotation in case anyone relies on them.

For now, can we copy/paste the description annotation and just have duplicated strings (we can do this after all editorial comments are dealt with).

We can deal with the actual CRD generation in step (2) later.

Reply:
I've made an issue (open-policy-agent/frameworks#140) requesting better support for `kubectl explain`.

A manual test I did suggests development work will be needed to get the description plumbed through:

```
❯ kubectl delete constrainttemplates k8srequiredlabels
constrainttemplate.templates.gatekeeper.sh "k8srequiredlabels" deleted

# APPLIES THE FOLLOWING:
#
# apiVersion: templates.gatekeeper.sh/v1beta1
# kind: ConstraintTemplate
# metadata:
#   name: k8srequiredlabels
# spec:
#   crd:
#     spec:
#       names:
#         kind: K8sRequiredLabels
#       validation:
#         # Schema for the `parameters` field
#         openAPIV3Schema:
#           description: "CARNE ASADA BURRITO"
#           properties:
#             labels:
#               type: array
#               items:
#                 type: string
#   targets:
#     - target: admission.k8s.gatekeeper.sh
#       rego: |
#         package k8srequiredlabels
# 
#         violation[{"msg": msg, "details": {"missing_labels": missing}}] {
#           provided := {label | input.review.object.metadata.labels[label]}
#           required := {label | label := input.parameters.labels[_]}
#           missing := required - provided
#           count(missing) > 0
#           msg := sprintf("you must provide labels: %v", [missing])
#         }
❯ kubectl apply -f example/templates/k8srequiredlabels_template.yaml
constrainttemplate.templates.gatekeeper.sh/k8srequiredlabels unchanged

❯ kubectl explain k8srequiredlabels
KIND:     K8sRequiredLabels
VERSION:  constraints.gatekeeper.sh/v1beta1

DESCRIPTION:
     <empty>

FIELDS:
   apiVersion   <string>
     APIVersion defines the versioned schema of this representation of an
     object. Servers should convert recognized schemas to the latest internal
     value, and may reject unrecognized values. More info:
     https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources

   kind <string>
     Kind is a string value representing the REST resource this object
     represents. Servers may infer this from the endpoint the client submits
     requests to. Cannot be updated. In CamelCase. More info:
     https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds

   metadata     <Object>
     Standard object's metadata. More info:
     https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata

   spec <Object>

   status       <>
```

I'll leave this Action Item for myself once I'm done with prose edits:

- [ ] Copy all `metadata.annotations.description` values to `spec.crd.spec.validation.openAPIV3Schema.description`

Reply:
LGTM